### ISSUE TITLE
design discussion

### ISSUE THREAD

**Issue by chessai** [mckeankylej/unpacked-maybe#3]
we could use a type family to make sure users can't nest Maybe's, but then we can't write a functor instance

or we could do the following, using unboxed sums:

data Maybe a = Maybe (# (# #) | a #)

even though we don't have unlifted newtypes, this is ok since the unboxing of this maybe type is trivial enough for GHC to perform without it

with this approach, we have to use 2 machine words instead of 1, but at least they're contiguous

**Comment by klapaucius** [comment]
> but at least they're contiguous

no, they are not

> The goal of this library was to cut down on the pointer indirections. So the unboxed sum trick would work.

no, it would not

```
newtype Foo a = Foo Any
Foo Int
+---+   +---+---+
|ptr|-->|inf|i64|
+---+   +---+---+
data Foo a = Foo (# (##) | a #)
Foo Int
+---+   +---+---+---+   +---+---+
|ptr|-->|inf|tag|ptr|-->|inf|i64|
+---+   +---+---+---+   +---+---+
-----------------------------------
+---+
|   | == 1 word
+---+
```

**Comment by chessai** [comment]
i think it would be useful for me to clarify about the second statement; "The goal of this library was to cut down on the pointer indirections. So the unboxed sum trick would work". This is in comparison to `Prelude.Maybe`, not `Any`. In this sense, the "unboxed sum trick" _does_ work.

Additionally, switching away from `Any` had to do with problems with nested `Data.Maybe.Unpacked.Maybe` values. This statement just explains the mention.

Also, your diagram and reasoning doesn't account for optimisations. I think you would be correct if no optimisations occurred. There are two main optimisations that we care about here: One, the core-to-core constructor elimination via worker-wrapper (e.g. in `data Foo a = Foo (# (# #) | a #)`, you can eliminate `Foo` and just have the unboxed tuple). Two, unarising. First, let's lay out roughly how an unboxed sum is represented, just so we're on the same page. Based on your inclusion of `tag` I can guess that you're already familiar with this, but for the sake of clarity I'm going to include it anyway:

Unboxed Sums are basically Unboxed Tuples, with an extra tag field that tells you which element is "active". Consider some primitive (i'm making up the name, its just a machine integer) for this tag called `Tag#`. Some examples:

```
(# Int# | Int# | Int# | ByteArray# )
```
would be represented like
```
(# Tag#, Int#, ByteArray# #)
```

One more:

```
(# Int# | (# Int#, Int# #) | Int# #)
```
would become something like
```
(# Tag#, Int#, Int# #)
```

Note that they shrink because of duplicate field folding, since they are never present simultaneously.

Consider the following module:
```
{-# language MagicHash #-}
{-# language UnboxedTuples #-}
{-# language UnboxedSums #-}
{-# OPTIONS_GHC -fforce-recomp -ddump-to-file -ddump-simpl -ddump-stg -ddump-cmm #-}
module Sums (Foo(..)) where
data Foo a = Foo (# (# #) | a #)
```

Now, consider the unarising that occurs. After unarising, `Foo` is basically `data Foo a = Foo Tag# a`, i.e. the payload of the datatype is just a tag and a pointer.

In STG, we can see the unarising that has occurred:
```
Before Unarise:
Sums.Foo :: forall a. (# (# #) | a #) -> Sums.Foo a
[GblId[DataCon],
 Arity=1,
 Caf=NoCafRefs,
 Str=<L,U>m,
 Unf=OtherCon []] =
    \r [eta_B1] Sums.Foo [eta_B1];
After Unarise:
Sums.Foo :: forall a. (# (# #) | a #) -> Sums.Foo a
[GblId[DataCon],
 Arity=1,
 Caf=NoCafRefs,
 Str=<L,U>m,
 Unf=OtherCon []] =
    \r [us_g1av us_g1aw] Sums.Foo [us_g1av us_g1aw];
```

Relevant portion of CMM that is actually writing a `Foo` to memory:
```
          I64[Hp - 16] = Sums.Foo_con_info;
           P64[Hp - 8] = _g1aw::P64;
           I64[Hp] = _g1av::I64;
           _c1aJ::P64 = Hp - 15;
           R1 = _c1aJ::P64;
           call (P64[Sp])(R1) args: 8, res: 0, upd: 8;
```

The CMM is really telling. Note particularly the contiguous writes of a `P64` (64-bit pointer) and an `I64` (64-bit integer)

TLDR: i think what you said makes sense, but not usually if -O2 is enabled, which im sure will be on if someone cares enough to use this library.
